server/src/main/java/org/spine3/server/storage/memory/InMemoryRecordStorage.java

```diff
@@ -0,0 +1,173 @@
+import org.spine3.protobuf.AnyPacker;
+import org.spine3.protobuf.TypeUrl;
+import org.spine3.server.entity.FieldMasks;
+import org.spine3.server.storage.EntityStorageRecord;
+import org.spine3.server.storage.RecordStorage;
+import org.spine3.server.users.CurrentTenant;
+import org.spine3.users.TenantId;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Maps.newHashMap;
+
+/**
+ * Memory-based implementation of {@link RecordStorage}.
+ *
+ * @author Alexander Litus, Alex Tymchenko
+ */
+/* package */ class InMemoryRecordStorage<I> extends RecordStorage<I> {
+
+    /** A stub instance of {@code TenantId} to be used by the storage in single-tenant context. */
+    private static final TenantId singleTenant = TenantId.newBuilder()
+                                                         .setValue("SINGLE_TENANT")
+                                                         .build();
+
+    private final Map<TenantId, Map<I, EntityStorageRecord>> tenantToStorageMap = newHashMap();
+
+    protected InMemoryRecordStorage(boolean multitenant) {
+        super(multitenant);
+    }
+
+    @SuppressWarnings("MethodWithMultipleLoops")    /* It's OK for in-memory implementation
+                                                     * as it is used primarily in tests. */
+    @Override
+    protected Iterable<EntityStorageRecord> readBulkInternal(final Iterable<I> givenIds, FieldMask fieldMask) {
+        final Map<I, EntityStorageRecord> storage = getStorage();
+
+        // It is not possible to return an immutable collection, since {@code null} may be present in it.
+        final Collection<EntityStorageRecord> result = new LinkedList<>();
+
+        TypeUrl typeUrl = null;
+
+        for (I recordId : storage.keySet()) {
+            for (I givenId : givenIds) {
+                if (recordId.equals(givenId)) {
+                    final EntityStorageRecord.Builder matchingRecord = storage.get(recordId).toBuilder();
+                    final Any state = matchingRecord.getState();
+
+                    if (typeUrl == null) {
+                        typeUrl = TypeUrl.of(state.getTypeUrl());
+                    }
+
+                    final Message wholeState = AnyPacker.unpack(state);
+                    final Message maskedState = FieldMasks.applyIfValid(fieldMask, wholeState, typeUrl);
+                    final Any processed = AnyPacker.pack(maskedState);
+
+                    matchingRecord.setState(processed);
+
+                    result.add(matchingRecord.build());
+                } else {
+                    result.add(null);
+                }
+            }
+        }
+        return result;
+    }
+
+    @Override
+    protected Iterable<EntityStorageRecord> readBulkInternal(Iterable<I> ids) {
+        return readBulkInternal(ids, null);
```

Comment:
The `fieldMask` parameter of `readBulkInternal(final Iterable<I> givenIds, FieldMask fieldMask)` is not annotated as `@Nullable`. Can we really pass `null`?